### ISSUE TITLE
[NbConvertApp] WARNING | Config option `codefolding` not recognized by `NbConvertApp`

### ISSUE THREAD

**Issue by lucainnocenti** [ipython-contrib/jupyter_contrib_nbextensions#1047]
I'm trying to export a notebook containing some `code_folding` into html, following [the relevant section in the guide](http://jupyter-contrib-nbextensions.readthedocs.io/en/latest/exporting.html#retaining-codefolding).

However, when running `jupyter nbconvert --to html --NbConvertApp.codefolding=True mynotebook.ipynb` the `codefolding` option seems to not be recognized, as per the title (I get 4 identical warnings about this), and the output html does not contain any folding.

I haven't tried the other option mentioned in the guide, that is, to modify the `jupyter_nbconvert_config.py` file, mainly because that file does not already exist in my system, while the guide seems to assume that it does and already contains some other code.
I have the `json` version of the file though, with the following content:
```
{
  "version": 1,
  "Exporter": {
    "template_path": [
      ".",
      "/home/lk/anaconda3/lib/python3.6/site-packages/jupyter_contrib_nbextensions/templates"
    ],
    "preprocessors": [
      "jupyter_contrib_nbextensions.nbconvert_support.CodeFoldingPreprocessor",
      "jupyter_contrib_nbextensions.nbconvert_support.PyMarkdownPreprocessor"
    ]
  }
}
```
So it looks like the exporting of the code folding should be already enabled, but for some reason it doesn't work.

**Comment by jcb91** [comment]
> `[NbConvertApp] WARNING | Config option 'codefolding' not recognized by 'NbConvertApp'

This happens because we're using a value that isn't provided by the `NbConvertApp` class. I guess really we should use a config value on the preprocessor class, so something like `CodeFoldingPreprocessor.codefolding`. I guess this is the source of the problem, since it seems that if the warning has been triggered, the value doesn't actually get set (see [traitlets/config/configurable.py#L162-L186](https://github.com/ipython/traitlets/blob/e9acdb51268ec3637d4054cfa1c261e3802e9455/traitlets/config/configurable.py#L162-L186)), which becomes a problem (see below).

> I haven't tried the other option mentioned in the guide, that is, to modify the `jupyter_nbconvert_config.py` file, mainly because that file does not already exist in my system, while the guide seems to assume that it does and already contains some other code.
I have the `json` version of the file though, with the following content:

Yes, the guide should probably link to [some information about jupyter configuration](http://jupyter-contrib-nbextensions.readthedocs.io/en/latest/config.html), which in general can be done either on the command line, using arguments like `--NbconvertApp.codefolding=True`, or in files, either `json` (preferred) or `.py` scripts (feel free to submit PRs to improve the docs!). The config you list shows that the preprocessor will be enabled; __however__, *the preprocessor doesn't actually do anything unless the `NbconvertApp.codefolding` key is __also__ set*, which is why the guide mentions setting either the command-line argument, or editing a config file. It should have noted that you can equivalently edit the json file, though.

**Comment by lucainnocenti** [comment]
Thanks for the useful info.
I will gladly contribute to clarify the docs, if I manage to understand the problem well enough.
However, I'm still not very clear of what the issue is here.
Is this just not working at the moment, or did I miss to do something?
You mention using a config value like `CodeFoldingPreprocessor.codefolding`, but where should this be changed?

**Comment by jcb91** [comment]
≥ I will gladly contribute to clarify the docs, if I manage to understand the problem well enough.

Thanks :smile:

Apologies, my explanation wasn't very clear.

> Is this just not working at the moment, or did I miss to do something?

I _think_ that this is currently broken, because of the attempt to use an unrecognised trait.

> You mention using a config value like `CodeFoldingPreprocessor.codefolding`, but where should this be changed?

This was me thinking about how the bug might get fixed, by using a config value belonging to the preprocessor (which we can add). Once implemented, it would be used in the same way as the existing docs indicate - either as a command line argument, or in a json/python config file.

I think the reason we've added the preprocessor as part of the install without enabling it is to make it easier to enable (you don't have to type the fully-qualified class name on the command line) without actually altering peoples' exports unrequested (you have to explicitly set the codefolding config value). Unfortunately, so far this seems not to work as intended...

As a temporary fix, you could edit the preprocessor python source to assume codefolding should be included.

Hope that makes a little more sense?

**Comment by jfbercher** [comment]
@lucainnocenti 
In your `jupyter_nbconvert_config.json` file, try to change the line 
```
      "jupyter_contrib_nbextensions.nbconvert_support.CodeFoldingPreprocessor",
```
into 
```
      "jupyter_contrib_nbextensions.nbconvert_support.pre_codefolding.CodeFoldingPreprocessor",
```
Then run nbconvert as
```
jupyter nbconvert --log-level='DEBUG' your_file.ipynb 
```
to see if the preprocessor is effectively called.

**Comment by lucainnocenti** [comment]
@jfbercher thanks for the tip.
With that change the warning is gone, but still the preprocessor is not called, and the folding not included in the output. Here is the relevant bit of the output:

```
[NbConvertApp] Converting notebook Untitled.ipynb to html
[NbConvertApp] Notebook name is 'Untitled'
[NbConvertApp] Applying preprocessor: coalesce_streams
[NbConvertApp] Applying preprocessor: CSSHTMLHeaderPreprocessor
[NbConvertApp] Applying preprocessor: HighlightMagicsPreprocessor
[NbConvertApp] Attempting to load template full
[NbConvertApp]     template_path: .
[NbConvertApp] Writing 312269 bytes to Untitled.html
```

**Comment by jcb91** [comment]
@lucainnocenti, it looks like your config file isn't being loaded, since there's no mention of an attempt to load the preprocessors specified. However, there's still a problem, since my run (on Ubuntu 16.04) looks like this:

```
$ source activate jup # activate correct env

$ jupyter nbconvert --debug --NbConvertApp.codef
olding=True --to html Berlina.ipynb              
[NbConvertApp] WARNING | Config option `codefoldi
ng` not recognized by `NbConvertApp`.            
[NbConvertApp] Searching ['/home/ubuntu/workspace
/Notebooks', '/home/ubuntu/.jupyter', '/home/ubun
tu/miniconda3/envs/jup/etc/jupyter', '/usr/local/
etc/jupyter', '/etc/jupyter'] for config files   
[NbConvertApp] Looking for jupyter_config in /etc
/jupyter                                         
[NbConvertApp] Looking for jupyter_config in /usr
/local/etc/jupyter                               
[NbConvertApp] Looking for jupyter_config in /hom
e/ubuntu/miniconda3/envs/jup/etc/jupyter         
[NbConvertApp] Looking for jupyter_config in /hom
e/ubuntu/.jupyter                                
[NbConvertApp] Looking for jupyter_config in /hom
e/ubuntu/workspace/Notebooks                     
[NbConvertApp] WARNING | Config option `codefoldi
ng` not recognized by `NbConvertApp`.            
[NbConvertApp] Looking for jupyter_nbconvert_conf
ig in /etc/jupyter                               
[NbConvertApp] Looking for jupyter_nbconvert_conf
ig in /usr/local/etc/jupyter                     
[NbConvertApp] Looking for jupyter_nbconvert_conf
ig in /home/ubuntu/miniconda3/envs/jup/etc/jupyte
r                                                
[NbConvertApp] Loaded config file: /home/ubuntu/m
iniconda3/envs/jup/etc/jupyter/jupyter_nbconvert_
config.json                                      
[NbConvertApp] Looking for jupyter_nbconvert_conf
ig in /home/ubuntu/.jupyter                      
[NbConvertApp] Looking for jupyter_nbconvert_conf
ig in /home/ubuntu/workspace/Notebooks           
[NbConvertApp] WARNING | Config option `codefoldi
ng` not recognized by `NbConvertApp`.            
[NbConvertApp] WARNING | Config option `codefoldi
ng` not recognized by `NbConvertApp`.            
[NbConvertApp] Converting notebook Berlina.ipynb 
to html                                          
[NbConvertApp] Notebook name is 'Berlina'        
[NbConvertApp] Applying preprocessor: coalesce_st
reams                                            
[NbConvertApp] Applying preprocessor: CSSHTMLHead
erPreprocessor                                   
[NbConvertApp] Applying preprocessor: HighlightMa
gicsPreprocessor                                 
[NbConvertApp] Applying preprocessor: CodeFolding
Preprocessor                                     
[NbConvertApp] Applying preprocessor: PyMarkdownP
reprocessor                                      
[NbConvertApp] Attempting to load template full  
[NbConvertApp]     template_path: .:/home/ubuntu/
workspace/jupyter_contrib_nbextensions/src/jupyte
r_contrib_nbextensions/templates                 
[NbConvertApp] Writing 251675 bytes to Berlina.ht
ml

$ grep -c -i -e "codefolding" -- Berlina.html # count occurrences of codefolding in output:
0
```

**Comment by jfbercher** [comment]
Haha. This is mysterious since it worked for me...?
```
[NbConvertApp] Applying preprocessor: coalesce_streams
[NbConvertApp] Applying preprocessor: CSSHTMLHeaderPreprocessor
[NbConvertApp] Applying preprocessor: HighlightMagicsPreprocessor
[NbConvertApp] Applying preprocessor: CodeFoldingPreprocessor
```

Other attempt: create a `codefolding_cfg.py` file with content
```python
import os

import jupyter_contrib_nbextensions.nbconvert_support

jcnbe_dir = os.path.dirname(jupyter_contrib_nbextensions.__file__)
pp_mod_name = 'jupyter_contrib_nbextensions.nbconvert_support.pre_codefolding'

c = get_config()  # noqa
c.NbConvertApp.export_format = "html"
c.Exporter.template_path = [
    '.',
    jupyter_contrib_nbextensions.nbconvert_support.templates_directory()
]
c.Exporter.preprocessors = [pp_mod_name + '.CodeFoldingPreprocessor']
```

and run 
```
jupyter nbconvert --log-level='DEBUG' --config ./codefolding_cfg.py your_file.ipynb 
```

**Comment by jcb91** [comment]
Apologies, confusion on my part, I forgot that codefolding isn't embedded, just the folded elements cut out. False alarm! @lucainnocenti it looks as though for some reason either your config is not being read, or it is being overridden by a later config...

**Comment by lucainnocenti** [comment]
well this is weird. 

First of all, about the config file, yes it's definitely not being loaded, but I'm not sure way. It's the only 'jupyter_nbconvert_config.json` file in my system, and it's definitely in one of the config directories. There are no `jupyter_nbconvert_config.py` around either.
I also tried to reference it directly with the `--config` option but to no avail, it is still not loaded.

I've tried to look for the line that handles this config loading but couldn't find it. `Grep`ping for `Looking for` in the `nbconvert` and `jupyter_contrib_nbextensions` repos gives nothing.
Do you know where is this loading handled?

I also tried @jfbercher's suggestion. Using that `.py` file as configuation works, and the `preprocessor` is listed as loaded and no errors are thrown.
However, the folded lines in the original `ipynb` are still not folded in the produced `html`.

**Comment by jcb91** [comment]
> it's definitely not being loaded, but I'm not sure way ... and it is in one of the config directories

Which config directories are actually being checked? Other possibilities include json syntax errors...

> Do you know where is this loading handled?

The config loading is from the traitlets package

> I also tried @jfbercher's suggestion

Yeah, that script looks ok, but still doesn't set the required `NbConvertApp.codefolding` value  -
it needs a line like 

```python
NbConvertApp.codefolding = True
```

as previously discussed

**Comment by lucainnocenti** [comment]
Well here is the complete output, listing also the directories that are being checked:
```
lk@lk-pc:/mnt/c/Users/lk/Documents/mydir$ jupyter nbconvert --log-level='DEBUG' --config /home/lk/anaconda3/etc/jupyter/jupyter_nbconvert_config.json Untitled.ipynb
[NbConvertApp] Searching ['/mnt/c/Users/lk/Documents/mydir', '/home/lk/.jupyter', '/home/lk/anaconda3/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[NbConvertApp] Looking for jupyter_config in /etc/jupyter
[NbConvertApp] Looking for jupyter_config in /usr/local/etc/jupyter
[NbConvertApp] Looking for jupyter_config in /home/lk/anaconda3/etc/jupyter
[NbConvertApp] Looking for jupyter_config in /home/lk/.jupyter
[NbConvertApp] Looking for jupyter_config in /mnt/c/Users/lk/Documents/mydir
[NbConvertApp] Looking for jupyter_nbconvert_config in /home/lk/anaconda3/etc/jupyter
[NbConvertApp] Converting notebook Untitled.ipynb to html
[NbConvertApp] Notebook name is 'Untitled'
[NbConvertApp] Applying preprocessor: coalesce_streams
[NbConvertApp] Applying preprocessor: CSSHTMLHeaderPreprocessor
[NbConvertApp] Applying preprocessor: HighlightMagicsPreprocessor
[NbConvertApp] Attempting to load template full
[NbConvertApp]     template_path: .
[NbConvertApp] Writing 312269 bytes to Untitled.html
```
while the content of the `.json` file is the one I [reported above](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/issues/1047#issuecomment-323042009).

**Comment by jcb91** [comment]
So from that output, it looks suspiciously like `/home/lk/anaconda3/etc/jupyter/jupyter_nbconvert_config.json` doesn't actually exist, since you've got the `Looking for jupyter_nbconvert_config in /home/lk/anaconda3/etc/jupyter`, but nothing further... :confused:

**Comment by lucainnocenti** [comment]
It seems we are finally getting there!

Of course you were right @jcb91 in that the file didn't exist.....ish!
It turns out that the problem came from the fact that I'm using bash on ubuntu on windows.
I created without thinking the `.json` config file using Notepad++ from Windows.
This created a weird situation in which the file was clearly there as shown from explorer, but wasn't seen by bash, and therefore by python run from anaconda from ubuntu!
[Not surprisingly](https://blogs.msdn.microsoft.com/commandline/2016/11/17/do-not-change-linux-files-using-windows-apps-and-tools/), one thing you should never do is create/modify files inside the lxss from Windows.

Anyway, I created the file again from inside the ubuntu distribution and it is now loaded just fine, so here is what happens now:

Loading the config from the `json` file *with the original setting* (so without adding the `pre_codefolding` suggested by @jfbercher, loads without warnings. Adding on top of this the command line option `--NbConvertApp.codefolding=True` produces the original warning about `codefolding` not being recognized by `NbConvertApp`.
*However*, despite the warning, the preprocessor is loaded and *the final html correctly contains the folding*.

Using the `pre_codefolding` add to the `json` file I have to remove the `PyMarkdownPreprocessor` line that was there by default, which causes an error when `traitlets` tries to import the corresponding module and doesn't find it (probably because I don't have the PyMarkdown extension installed?).
Once this line is removed, the behaviour is the exact same as above, so it looks like `pre_codefolding` isn't doing much?

Using the `.py` configuration file is similar: adding the line `c.NbConvertApp.codefolding = True` and running `nbconvert` (this time without needing to add the option of course) produces the usual warning *and* the correct output html.

So at the end of the day the warning is annoying but not really the problem.
The exporting wasn't working for me simply because the config file was not read.
Do you know how to get rid of that?

Also, on a different note: I noticed that in the output `html` the folded code is just removed, and it is not possible to do the dynamic folding anymore.
Is there any reason for this? Would it be possible to add via js the same kind of dynamic folding that we have in the notebook?

**Comment by jfbercher** [comment]
> Another question now arises: why does it work? As pointed out by @jcb91 before, if the warning is issued the property is not set... but it looks like it is anyway?

I think it is set *after* the warning is issued. Probably that nbconvertApp first checks if `codefolding` is among  awaited traits and it is not. Then this value is added as a key of the NbConvertApp object and that value is tested in `pre_codefolding`. 
I just checked that changing [pre_cofolding.py#L72](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/blob/master/src/jupyter_contrib_nbextensions/nbconvert_support/pre_codefolding.py#L72) into
```
dofolding = self.config.CodeFolding.get('codefolding', False) is True
```
would enable to do something like
```
jupyter nbconvert --CodeFolding.codefolding=True FILE.ipynb
```
without warning.

**Comment by jfbercher** [comment]
@juhasch Thoughts on the possible modification mentioned above?

**Comment by jfbercher** [comment]
> Also, on a different note: I noticed that in the output html the folded code is just removed, and it is not possible to do the dynamic folding anymore. Is there any reason for this? Would it be possible to add via js the same kind of dynamic folding that we have in the notebook?

As far as I see, the `codefolding` nbextension relies on enabling the codemirror codefolding feature. This one is not present in the html exported version; so this cannot be done that way. 

It could probably be possible to embed the folded lines in a div and add a class such that visibility could be toggled with a button/link/caret and some lines of js.

**Comment by jcb91** [comment]
Aha! Good detective work @lucainnocenti :) I think I've run into a similar problem before, but forgotten all about it since ditching Windows haha. Anyway, glad you've got to the bottom of it.

> So at the end of the day the warning is annoying but not really the problem,
and the exporting wasn't working for me simply because the config file was not read.
Do you know how to get rid of the warning?

Essentially we should use a config value that either exists (because we defined it on the relevant class) or is attached to a nonexistent class, in order not to produce warnings. I'd prefer to use a value on the preprocessor, as I mentioned previously.

> Also, on a different note: I noticed that in the output `html` the folded code is just removed, and it is not possible to do the dynamic folding anymore.
Is there any reason for this? Would it be possible to add via js the same kind of dynamic folding that we have in the notebook?

The main reason is that nobody's had time to implement it yet :wink: however, the folding might need more effort to implement, as currently the frontend relies on codemirror, which I don't think is included in html output...

**Comment by jcb91** [comment]
Perhaps we ought to use `Preprocessor.enabled`, as in https://github.com/jupyter/nbconvert/blob/68a3925efef5c304c2c8682ad05cac9560c4173a/nbconvert/preprocessors/base.py#L27

**Comment by juhasch** [comment]
@jfbercher: I think the idea of @jcb91 sounds good. The original `--NbConvertApp.codefolding=True` never was a good solution.

@lucainnocenti: As @jcb91 already said, the folded code is removed because otherwise it would mean either using codemirror in the html output or reimplementing the folding logic.
Maybe there should be two html types nbconvert can generate : 'static' for documenting and 'live' to display a page with the Javascript like codefolding and plots stuff still working.

The preprocessor part in your JSON file should look like this:
```
    "preprocessors": [
      "jupyter_contrib_nbextensions.nbconvert_support.CodeFoldingPreprocessor",
      "jupyter_contrib_nbextensions.nbconvert_support.PyMarkdownPreprocessor"
    ]
```
This is because in nbconvert_support/__init__.py all nbconvert Preprocessors are imported.

**Comment by jfbercher** [comment]
> Perhaps we ought to use Preprocessor.enabled, as in 

I am afraid this will not work. After investigating a bit, the preprocessor list we provide in the JSON config file correspond to the trait in [exporter.py#L63-L66](https://github.com/jupyter/nbconvert/blob/68a3925efef5c304c2c8682ad05cac9560c4173a/nbconvert/exporters/exporter.py#L63-L66) and these exporters are automatically enabled [exporter.py#L249-L251](https://github.com/jupyter/nbconvert/blob/68a3925efef5c304c2c8682ad05cac9560c4173a/nbconvert/exporters/exporter.py#L249-L251). Do I miss something?
The enabled flag seems to be used to enable one of the default processor (which may not be enabled by default). 
__

As suggested above by @jcb91, we can define a config value in the relevant class and use it. I checked that using a custom trait in `CodeFoldingPreprocessor` class
```
 codefolding = Bool(False, shortname="rh",
                         help="do codefolding or not").tag(config=True)
```
with a `preprocess` method 
```
def preprocess(self, nb, resources):

        if self.codefolding: 
            return super(CodeFoldingPreprocessor, self).preprocess(nb, resources)
        else:
            #print("Applying CodeFoldingPreprocessor: disabled")
            return nb, resources 
```
enables CodeFolding preprocessing with
`jupyter nbconvert  --CodeFoldingPreprocessor.codefolding=True FILE.ipynb`
otherwise the preprocessing is not done.

**Comment by jcb91** [comment]
Ah, yes, you're absolutely right, I had a vague memory of something like this with the enabled trait and couldn't quite recall it, but you've got it right, it gets set by the exporter so not useful for this issue.

Custom trait sounds like the way forwards :)

**Comment by jfbercher** [comment]
> Though perhaps we could pick a more descriptive name than just 'codefolding'? Something like 'remove_folded_code'? Not sure what makes most sense

Yes, that makes sense I think. It leaves the possibility, someday, to add another option to enable embedding js for 'dynamic' codefolding.

**Comment by juhasch** [comment]
Sounds like a good idea to have an `active` export that retains codemirror features and a `passive` export meant for documentation. I wonder how difficult it would be just to cut the kernel connection of a live notebook ?

**Comment by jcb91** [comment]
> I wonder how difficult it would be just to cut the kernel connection of a live notebook ?

I'm not sure what you mean by this? Sort of like exporting *all* the notebook frontend, minus kernel connections?

**Comment by juhasch** [comment]
I was just thinking, if you had a read-only notebook, what do you actually need a kernel connection for (except widgets). Not saying it really makes sense. On the other hand, things like toc2, codefolding etc. would work.
Maybe we could just monkey patch the kernel communication JS and just return OK without doing anything :smile:

**Comment by jcb91** [comment]
> Maybe we could just monkey patch the kernel communication JS and always return OK without doing anything :smile:

Haha this would probably work quite well!

**Comment by jcb91** [comment]
Should be fixed by #1058 (please reopen if not!)
